hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/TestFsDatasetImpl.java

```diff
@@ -1101,15 +1099,12 @@ public void testReportBadBlocks() throws Exception {
 
       block = DFSTestUtil.getFirstBlock(fs, filePath);
       // Test for the overloaded method reportBadBlocks
-      dataNode.reportBadBlocks(block, dataNode.getFSDataset()
-          .getFsVolumeReferences().get(0));
-      Thread.sleep(3000);
-      BlockManagerTestUtil.updateState(cluster.getNamesystem()
-          .getBlockManager());
-      // Verify the bad block has been reported to namenode
-      Assert.assertEquals(1, cluster.getNamesystem().getCorruptReplicaBlocks());
-    } finally {
-      cluster.shutdown();
+      dataNode.reportBadBlocks(block, dataNode.getFSDataset().getFsVolumeReferences().get(0));
+      GenericTestUtils.waitFor(() -> {
+        BlockManagerTestUtil.updateState(cluster.getNamesystem().getBlockManager());
+        // Verify the bad block has been reported to namenode
+        return 1 == cluster.getNamesystem().getCorruptReplicaBlocks();
+      }, 100, 10000, "Corrupted replica blocks could not be found");
```

Comment:
You are able to repro, right? Does rather than waiting, shooting Heartbeat works

Reply:
Oh that should be a better option actually

Reply:
With the same debug mode that I am able to repro original failure with, I am able to confirm that heartbeat works for sure. I just don't recall if we have also seen heartbeat specific flakes too in Jenkins.

Reply:
Basically what I am trying to say is that whether we should also consider increasing wait time here, by say 500/1000 ms instead of 100 ms?

```
  void triggerHeartbeatForTests() {
    synchronized (ibrManager) {
      final long nextHeartbeatTime = scheduler.scheduleHeartbeat();
      ibrManager.notifyAll();
      while (nextHeartbeatTime - scheduler.nextHeartbeatTime >= 0) {
        try {
          ibrManager.wait(100);
        } catch (InterruptedException e) {
          return;
        }
      }
    }
  }

```

Reply:
I think triggering a heartbeat is not directly related to `reportBadBlocks`? BadBlocks are processed in `BPServiceActor#processQueueMessages`. What do you think of this？

Reply:
You are correct in the sense that it is not heartbeat trigger as such that makes namenode aware of the bad block, but it is the trigger of `ReportBadBlockAction#reportTo` that calls `DatanodeProtocol#reportBadBlocks` and hence Namenode gets notified. All this happens as part of `BPServiceActor#processQueueMessages`, you are right.

However, it takes place immediately after sending heartbeat, as part of the same `offerService` call. Hence, the point here is, as long as the main thread (test thread) is waiting for enough time to let this whole round of `offerService` to complete, we should be good. This is exactly what `DataNodeTestUtils#triggerHeartbeat` takes care of, perhaps there might be some room for improvement as mentioned above that we could increase extra wait time from 100ms to maybe a bit higher, but I am sure this utility is used by multiple tests and the wait time would have been sufficient so far (as no one has attempted to change it so far).

But as to what you mentioned, it's correct. The main purpose however is for us to let `offerService` complete it's whole run and we wait by then as part of `DataNodeTestUtils#triggerHeartbeat`, hence I am not able to reproduce flaky behavior with this change, at least locally. Does this sound good to you @tomscut?

Reply:
I gave it a try and was able to repro this. With **triggerHeartbeat**, it worked for me. I think that is a standard practice for such cases running since legacy time for such cases(At least my time). Viraj I tried to reduce the ibrManager.wait(100) to even 1 and it still passes for me. I couldn't repro the case you are mentioning, do you intend to say if we put a sleep just before processQueueMessages, things should screw up? 

Well touching that util for our use case won't be a good idea. If it works for others, the intent of that util is to just make sure the heartbeat got sent and if it does that correctly, means doesn't return before sending heartbeat, it is good enough to stay as is. If something breaks then it is on the guy using it, can't blame this util.

If odds are against triggerHeartbeat, I would still try to avoid this sleep, Is something like this an option
```
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPOfferService.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPOfferService.java
index fdd66cb05d3..1dc86adafc8 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPOfferService.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPOfferService.java
@@ -76,6 +76,10 @@
   private volatile String bpId;
   private final DataNode dn;
 
+  BPServiceActor getBpServiceToActive() {
+    return bpServiceToActive;
+  }
+
   /**
    * A reference to the BPServiceActor associated with the currently
    * ACTIVE NN. In the case that all NameNodes are in STANDBY mode,
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPServiceActor.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPServiceActor.java
index e9f424604b4..456826f630c 100755
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPServiceActor.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPServiceActor.java
@@ -1001,7 +1001,8 @@ public void bpThreadEnqueue(BPServiceActorAction action) {
     }
   }
 
-  private void processQueueMessages() {
+  @VisibleForTesting
+  void processQueueMessages() {
     LinkedList<BPServiceActorAction> duplicateQueue;
     synchronized (bpThreadQueue) {
       duplicateQueue = new LinkedList<BPServiceActorAction>(bpThreadQueue);
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/DataNodeTestUtils.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/DataNodeTestUtils.java
index dba5a146f0c..b70d9afd30e 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/DataNodeTestUtils.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/DataNodeTestUtils.java
@@ -269,4 +269,13 @@ public Boolean get() {
       }
     }, 100, 10000);
   }
+
+  /**
+   * Processes all the enqueued messages to active namenode
+   * @param dataNode the datanode.
+   */
+  public static void processQueueMessagesToActive(DataNode dataNode) {
+    dataNode.getAllBpOs().forEach(bpos -> bpos.getBpServiceToActive().processQueueMessages());
+  }
+
 }
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/TestFsDatasetImpl.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/TestFsDatasetImpl.java
index d6f42f3d020..a4d400f2777 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/TestFsDatasetImpl.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/TestFsDatasetImpl.java
@@ -1101,9 +1101,8 @@ public void testReportBadBlocks() throws Exception {
 
       block = DFSTestUtil.getFirstBlock(fs, filePath);
       // Test for the overloaded method reportBadBlocks
-      dataNode.reportBadBlocks(block, dataNode.getFSDataset()
-          .getFsVolumeReferences().get(0));
-      Thread.sleep(3000);
+      dataNode.reportBadBlocks(block, dataNode.getFSDataset().getFsVolumeReferences().get(0));
+      DataNodeTestUtils.processQueueMessagesToActive(dataNode);
       BlockManagerTestUtil.updateState(cluster.getNamesystem()
           .getBlockManager());
       // Verify the bad block has been reported to namenode

```

Reply:
> I gave it a try and was able to repro this. With triggerHeartbeat, it worked for me. I think that is a standard practice for such cases running since legacy time for such cases(At least my time).

Great, sound good.

> do you intend to say if we put a sleep just before processQueueMessages, things should screw up?

Yes that's what I was suspecting so far but looks like even with sleep, reproducing the screw up is difficult. I tried it now several times but unable to reproduce failure with sleep.

Also, now I understand what could still possibly go wrong with the race condition, let me update the patch.

Reply:
Let me create follow-up work to make this strongly consistent with reporting of the bad block.